### ISSUE TITLE
AutoRest tracing with xUnit v2 and test parallelization

### ISSUE THREAD

**Issue by gitfool** [Azure/autorest#1003]
**Scenario**:
Running _integration_ tests via xUnit, where these tests are using an AutoRest generated proxy to interact with a server, logging should be enabled for all output including from AutoRest, and these tests should run in parallel where possible.

Wire up a bridge between AutoRest tracing and xUnit output:
```c#
public sealed class XunitTracingInterceptor : IServiceClientTracingInterceptor, IDisposable
{
    public XunitTracingInterceptor(ITestOutputHelper logger)
    {
        _logger = logger;
        ServiceClientTracing.IsEnabled = true;
        ServiceClientTracing.AddTracingInterceptor(this);
    }

    public void Dispose()
    {
        ServiceClientTracing.RemoveTracingInterceptor(this);
    }

    public void Configuration(string source, string name, string value)
    {
    }

    public void EnterMethod(string invocationId, object instance, string method, IDictionary<string, object> parameters)
    {
        _logger.WriteLine($">>>> {method}");
    }

    public void ExitMethod(string invocationId, object returnValue)
    {
        _logger.WriteLine("<<<<");
    }

    public void Information(string message)
    {
        _logger.WriteLine(message);
    }

    public void ReceiveResponse(string invocationId, HttpResponseMessage response)
    {
        _logger.WriteLine($"Response: {response.AsFormattedString()}");
    }

    public void SendRequest(string invocationId, HttpRequestMessage request)
    {
        _logger.WriteLine($"Request: {request.AsFormattedString()}");
    }

    public void TraceError(string invocationId, Exception exception)
    {
        _logger.WriteLine($"Error: {exception}");
    }

    private readonly ITestOutputHelper _logger;
}
```
Use bridge from an integration test (one of many such tests):
```c#
[Trait("Category", "Integration")]
public class ProxyTests : IDisposable
{
    public ProxyTests(ITestOutputHelper logger)
    {
        _xunitTracingInterceptor = new XunitTracingInterceptor(logger);
        _proxy = new TestService();
    }

    public void Dispose()
    {
        _xunitTracingInterceptor.Dispose();
    }

    [Fact]
    public async Task Test()
    {
        // use _proxy...
    }

    private readonly XunitTracingInterceptor _xunitTracingInterceptor;
    private readonly TestService _proxy;
}
```

**Problem**:
It looks this won't work since AutoRest is using statics via `ServiceClientTracing` and xUnit requires a separate instance of `ITestOutputHelper` for each parallelizable output. When multiple proxy tests are running in parallel, the AutoRest tracing will be sent to _all_ interceptors, rather than just the interceptor for a given proxy instance.

**References**:
* [AutoRest: Tracing](https://github.com/Azure/autorest/blob/master/Documentation/clients-tracing.md)
* [ServiceClientTracing.cs](https://github.com/Azure/autorest/blob/master/ClientRuntimes/CSharp/Microsoft.Rest.ClientRuntime/ServiceClientTracing.cs)
* [xUnit: Capturing Output](https://xunit.github.io/docs/capturing-output.html)
* [xUnit: Running Tests in Parallel](https://xunit.github.io/docs/running-tests-in-parallel.html)

**Comment by gitfool** [comment]
**Conclusion**:
Probably not such a big deal since the minimum granularity for parallelism is a class. To avoid the problem above, move tests using other proxies into their own test assembly, so they each get their own  statics.

**Comment by dennisroche** [comment]
Alternative is to implement logging using the `DelegatingHandler`. You can set the handlers per ServiceClient.

```csharp
public class XUnitHttpLoggingHandler : DelegatingHandler
{
    private readonly ITestOutputHelper _logger;
    
    public XUnitHttpLoggingHandler(ITestOutputHelper logger) 
    {
        _logger = logger;
    }

    protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
    {
        var timer = new Stopwatch();
        timer.Start();
        var response = await base.SendAsync(request, cancellationToken);
        timer.Stop();

        _logger.WriteLine(string.Format("HTTP {0} '{1}'. Response was {2} in {3} ms",
            response.RequestMessage.Method.Method,
            response.RequestMessage.RequestUri,
            response.StatusCode,
            timer.ElapsedMilliseconds));

        if (response.StatusCode != HttpStatusCode.OK)
        {
            _logger.WriteLine(string.Format("Request Failed: {0}", response.Content == null ? "No error in content" : await response.Content.ReadAsStringAsync()));
        }

        return response;
    }

}
```

**Comment by gitfool** [comment]
Thanks @dennisroche; good to have options! It would still be preferable for the _built-in_ logging facility to be injectable per proxy instance.

**Comment by fearthecowboy** [comment]
Howdy!

In our planning for driving towards a stable '1.0' release, I'm marking this issue as 'deferred' :zzz: and we're going to review it during the post-1.0 planning cycle. 

_It's not to say that we're not going to work on it, or that this isn't not important, but at the moment, we're picking and choosing the stuff we **must** do before 1.0._ :horse_racing:  :horse_racing:  :horse_racing: 

We'll make sure we pick this back up at that point. :tada:
